source/tutorials/installing-nixos-on-a-raspberry-pi.rst

```diff
@@ -0,0 +1,175 @@
+  $ unzstd -d nixos-sd-image-21.05pre288297.8eed0e20953-aarch64-linux.img.zst
+  $ dmesg -w
+
+Your terminal should be printing kernel messages as they come in.
+
+Plug in your SD card and your terminal should print what device it got assigned, for example ``/dev/sdX``.
+
+Press ``ctrl-c`` to stop ``dmesg -w``.
+
+Copy NixOS to your SD card by replacing ``sdX`` with the name of your device:
+
+.. code:: shell-session 
+
+  sudo dd if=nixos-sd-image-21.05pre288297.8eed0e20953-aarch64-linux.img of=/dev/sdX bs=4096 conv=fsync status=progress
+
+Once that command exits, move the SD card into your RPi and power it on.
+
+You should be greeted with a fresh shell!
+
+Run ``sudo su`` for the rest of the tutorial.
+
+
+Getting internet connection
+---------------------------
+
+At this point we'll need internet connection. If you can use an ethernet cable, plug it in.
+
+In case you're connecting to a wifi run ``iwconfig`` to see what is the name of your wireless
+network interface. In case it's ``wlan0`` replace ``SSID`` and ``passphrase`` with your data and run:
+
+.. code:: shell-session 
+
+  $ wpa_supplicant -B -i wlan0 -c <(wpa_passphrase 'SSID' 'passphrase') &
+
+
+Once you see in your terminal that connection is established, run ``host google.com`` to 
+check that DNS resolves correctly.
+
+In case you've made a typo, run ``pkill wpa_supplicant`` and start over.
+
+
+Updating firmware
+-----------------
+
+To increase chances for hardware to work, we'll start by updating RPi firmware:
+
+.. code:: shell-session
+
+  $ nix-shell -p raspberrypi-eeprom
+  $ FIRMWARE_RELEASE_STATUS=stable rpi-eeprom-update -d -a
+
+  
+Installing NixOS 
+----------------
+
+For initial installation we'll install `XFCE <https://www.xfce.org/>`_ desktop environment
+with user ``guest`` and SSH daemon.
+
+.. code:: nix 
+
+  { config, pkgs, lib, ... }:
+
+  let
+    user = "guest";
+    password = "guest";
+    SSID = "mywifi";
+    SSIDpassword = "mypassword";
+    interface = "wlan0";
+    hostname = "myhostname";
+  in {
+    imports = ["${fetchTarball "https://github.com/domenkozar/nixos-hardware/archive/rpi4.tar.gz" }/raspberry-pi/4"];
+
+    fileSystems = {
+      "/" = {
+        device = "/dev/disk/by-label/NIXOS_SD";
+        fsType = "ext4";
+        options = [ "noatime" ];
+      };
+    };
+
+    networking = {
+      hostName = hostname;
+      wireless = {
+        enable = true;
+        networks."${SSID}".psk = SSIDpassword;
+        interfaces = [ interface ];
+      };
+    };
+
+    environment.systemPackages = with pkgs; [ vim ];
+
+    services.openssh.enable = true;
+
+    users = {
+      mutableUsers = false;
+      users.guest = {
+        isNormalUser = true;
+        password = password;
+        extraGroups = [ "wheel" ];
+      };
+    };
+
+    # video
+    services.xserver = {
+      enable = true;
+      displayManager.lightdm.enable = true;
+      desktopManager.xfce.enable = true;
+      videoDrivers = [ "fbdev" ];
```

Comment:
"modesetting" should work, and better, I believe.

(I'll have to boot one of my devices to validate some parts of the config.)

Reply:
I've tried it and it fails to start, it's probably because that's supposed to use GPU, but that doesn't work yet - I have yet to open an issue for this.

Reply:
https://github.com/NixOS/nixos-hardware/pull/261

We can drop this line entirely, and instead add

```nix
{
  hardware.raspberry-pi."4".fkms-3d.enable = true;
}
```